### DISCUSSION TITLE
Sending alerts from behind proxy

### DISCUSSION THREAD

**Discussion by dipdill** [apache/skywalking#10309] [Q&A]
Is there a way to use a proxy setting with the alarm-settings.yml?  I want to send Slack Messages but we are behind a company proxy server. 
 
Thanks

**Comment by wu-sheng** [comment]
Are you asking for some auth setup? If so, there isn't yet.
Contribution is welcome.

**Comment by dipdill** [comment]
No, my Skywalking server sits behind a proxy server on our network.  The only way we can send messages to any address on the web... ie: slack.com we need to pass it through the proxy.  Is there a configuration somewhere that will allow me to set a proxy server?

**Reply by wu-sheng** [reply]
I can't see any difference between server addr or proxy. The key of being proxy is transparent.
I am not sure what you are asking.

**Comment by dipdill** [comment]
Yes I totally agree that proxy servers are transparent.  Whenever i need to access any resource on the internet i need to do a 

export HTTP_PROXY=http://10.10.16.16:80
export HTTPS_PROXY=http://10.10.16.16:80
curl https://github.com

or

curl -x 10.10.16.16:80 https://github.com

Right now the alerts are trying to send to https://hooks.slack.com and never make it outside our company network... because the rest call isn't configured to send the request through the proxy server.

Is there a configuration setting within skywalking that can be set to configure a proxy for outbound rest calls?

**Reply by wu-sheng** [reply]
You could use any gateway(supports these settings). I believe Nignx and Envoy could do this easily.

**Reply by wu-sheng** [reply]
> You could use any gateway(supports these settings). I believe Nignx and Envoy could do this easily.

This is the solution. Search for Nginx and Envoy.
A proxy is a gateway to add auth information.

Like I said, if you insist, change the codes on your private side. You don't need to persuade me.

Ask for helps to the ppl around you having experience. We as an open source community can never do such detailed things.

**Comment by dipdill** [comment]
Ok i see the gateways.yml file and i added

gateways:
  - name: proxy0
    instances:
      - host: 10.10.16.16 # the host/ip of this gateway instance
        port: 80 # the port of this gateway instance, defaults to 80

restarted the service... and the alarms are still not using the proxy server to send.  Do i need to make a change in the alarm-settings.yml to tell it to use this new gateway for slack outbound messages?

**Reply by wu-sheng** [reply]
I don't know how your gateway works. Just point slack address to gateway.
This is a general deployment setup, there is nothing particular for OAP.

**Reply by Superskyyy** [reply]
Gateway.yml is not for what you think it is. It's for instrumenting gateways without a plugin. https://skywalking.apache.org/docs/main/next/en/setup/backend/uninstrumented-gateways/#configuration-format

What you need to change is in your company proxy rules, not something inside skywalking. Only thing should be changed in SkyWalking config is the address of the proxy to replace the original alarm webhook url. If you don't have access to that proxy config, either setup a front to rewrite request or modify oap source code to add option to proxy alarm send requests.

**Reply by dipdill** [reply]
Ok so I remove the gateway proxy settings...

If I change the destination from hooks.slack.com to 10.10.16.16:80... my proxy server isn't going to know that the payload is destined for hooks.slack.com.  You need to specify a proxy server in the request to tell it to send to the proxy to be forwarded to hooks.slack.com.

**Comment by dipdill** [comment]
I dont have access to modify rules on the proxy server... That is maintained by corporate IT.  They kill all web based traffic at the firewalls unless it is going through the proxy server.  Its no different than having to configure a proxy in a web browser.  If I remove my proxy settings in Firefox/Chrome... I can not surf the web.  

So I guess my question is:  Is there a way to configure Skywalking to use a proxy for external communications?  IE: sending web rest messages to slack, msteams, wechat or anything else off prem.

If not... what would it take to implement that?

**Reply by wu-sheng** [reply]
I think you should reach people having basic infrastructure setup experience to read this thread. He will understand.
We had provided the solution, but you can't follow.

**Reply by dipdill** [reply]
I dont think you provided any solution on how to send messages through a proxy.

**Comment by dipdill** [comment]
Here is a perfect example of what i am looking for https://kodejava.org/how-do-i-setting-a-proxy-for-httpclient/

`public class HttpGetProxy {
    private static final String PROXY_HOST = "proxy.host.com";
    private static final int PROXY_PORT = 8080;

    public static void main(String[] args) {
        HttpClient client = new HttpClient();
        HttpMethod method = new GetMethod("https://kodejava.org");

        HostConfiguration config = client.getHostConfiguration();
        config.setProxy(PROXY_HOST, PROXY_PORT);

        String username = "guest";
        String password = "s3cr3t";
        Credentials credentials = new UsernamePasswordCredentials(username, password);
        AuthScope authScope = new AuthScope(PROXY_HOST, PROXY_PORT);

        client.getState().setProxyCredentials(authScope, credentials);

        try {
            client.executeMethod(method);

            if (method.getStatusCode() == HttpStatus.SC_OK) {
                String response = method.getResponseBodyAsString();
                System.out.println("Response = " + response);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            method.releaseConnection();
        }
    }
}`

**Reply by wu-sheng** [reply]
If you want to change codes, go ahead. Apache 2.0 license granted you. 
Meanwhile, we as upstream would not. Like I replied, the solution is clear provided.

**Comment by dipdill** [comment]
Me being a little more dim than most... here would be my proposed solution...

inside the alarm-settings.yml file... allow for an optional proxy settings under the webhooks... something like this.

slackHooks:
  textTemplate: |-
    {
        "type": "section",
        "text": {
          "type": "mrkdwn",
          "text": ":alarm_clock: *Apache Skywalking Alarm* \n **%s**."
        }
    }      
  webhooks:
    - https://hooks.slack.com/services/T/B/O
  proxy:
    - host: 10.10.16.16
      port: 80
      
 Or even maybe have a proxy.yml file.  I dont know.
  
Then in the AlarmRulesWatcher.java you can gather those settings... like you do with all the settings and then pass that along to the SlackhookCallback.java or any other call back java file... perform a simple null check against it to see if proxy is set... and if it does exist... add it to the build line

CloseableHttpClient httpClient = HttpClients.custom().build();

So it becomes 

CloseableHttpClient httpClient = HttpClients.custom().setProxy(httpProxy).build();

And then Skywalking can support simple proxy settings and make it compliant with growing security demands.  Instead of needing modify security rules on the proxy server and firewalls to allow this traffic to pass in the clear.

I'm no developer and i really like the skywalking tool so far.  This just seems like a better solution in my opinion.  Thoughts?
